### PR TITLE
refactor: Convert NoteEditor to Kotlin 

### PR DESCRIPTION
**Pull request by david-allison**
## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Review by Akshay0701 — COMMENTED** [review]
Big hit!

there are still many minor code styles changes that can be done like using not empty or using `?.` instead of if statement etc

are you planning to do these changes in your next/IDE-lint PR?

**Comment by david-allison** [comment]
I want to keep the conversion PR as 'pure' and close to automated as possible.

The plan is for someone (maybe/likely me) who will go through later and will fix the cleanups.

I typically like doing the cleanups in two stages: fixing all the IDE lint, then fixing the cleanup attributes.

I do this because once the IDE lint is fixed, it's a strong incentive to avoid introducing any more, and we'll eventually be able to automate all of the checks

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 1]
```diff
@@ -16,2224 +16,2011 @@
...
+    private fun openNewNoteEditor(intentEnricher: Consumer<Intent>) {
+        val intent = Intent(this@NoteEditor, NoteEditor::class.java)
+        intent.putExtra(EXTRA_CALLER, CALLER_CARDEDITOR)
+        intent.putExtra(EXTRA_DID, deckId)
+        // mutate event with additional properties
+        intentEnricher.accept(intent)
+        startActivityForResultWithAnimation(
+            intent,
+            REQUEST_ADD,
+            ActivityTransitionAnimation.Direction.START
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:1001]: import `com.ichi2.anim.ActivityTransitionAnimation.Direction.*` back

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 2]
```diff
@@ -16,2224 +16,2011 @@
...
         }
         // ensure there are no orphans from possible edit previews
-        TemporaryModel.clearTempModelFiles();
+        TemporaryModel.clearTempModelFiles()
         if (mCaller == CALLER_CARDEDITOR_INTENT_ADD) {
-            finishWithAnimation(NONE);
+            finishWithAnimation(ActivityTransitionAnimation.Direction.NONE)
         } else {
-            finishWithAnimation(END);
+            finishWithAnimation(ActivityTransitionAnimation.Direction.END)
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:1094]: ditto

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 3]
```diff
@@ -16,2224 +16,2011 @@
...
-
+            intent.putExtra("noteId", mCurrentEditedCard!!.note().id)
+            Timber.d("showCardTemplateEditor() with note %s", mCurrentEditedCard!!.note().id)
+            intent.putExtra("ordId", mCurrentEditedCard!!.ord)
+            Timber.d("showCardTemplateEditor() with ord %s", mCurrentEditedCard!!.ord)
+        }
+        startActivityForResultWithAnimation(
+            intent,
+            REQUEST_TEMPLATE_EDIT,
+            ActivityTransitionAnimation.Direction.START
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:1139]: ditto

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 4]
```diff
@@ -16,2224 +16,2011 @@
...
+            return
         }
         if (note == null || mAddNote || mCurrentEditedCard == null) {
-            Model model = getCol().getModels().current();
-            if (getCol().get_config("addToCur", true)) {
-                mCurrentDid = getCol().get_config_long(CURRENT_DECK);
-                if (getCol().getDecks().isDyn(mCurrentDid)) {
+            val model = col.models.current()
+            if (col.get_config("addToCur", true)!!) {
+                deckId = col.get_config_long(Decks.CURRENT_DECK)
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:1619]: import `com.ichi2.libanki.Decks.CURRENT_DECK`

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 5]
```diff
@@ -16,2224 +16,2011 @@
...
-        Intent intent = new Intent(this, CardTemplateEditor.class);
+    private fun showCardTemplateEditor() {
+        val intent = Intent(this, CardTemplateEditor::class.java)
         // Pass the model ID
-        intent.putExtra("modelId", getCurrentlySelectedModel().getLong("id"));
-        Timber.d("showCardTemplateEditor() for model %s", intent.getLongExtra("modelId", NOT_FOUND_NOTE_TYPE));
+        intent.putExtra("modelId", currentlySelectedModel!!.getLong("id"))
+        Timber.d(
+            "showCardTemplateEditor() for model %s",
+            intent.getLongExtra("modelId", Models.NOT_FOUND_NOTE_TYPE)
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:1127]: Import `com.ichi2.libanki.Models.NOT_FOUND_NOTE_TYPE`

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 6]
```diff
@@ -16,2224 +16,2011 @@
...
-    protected void onStop() {
-        super.onStop();
-        if (!isFinishing()) {
-            WidgetStatus.update(this);
-            UIUtils.saveCollectionInBackground();
+    override fun onStop() {
+        super.onStop()
+        if (!isFinishing) {
+            WidgetStatus.update(this)
+            saveCollectionInBackground()
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:547]: Keep `UIUtils.`

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 7]
```diff
@@ -16,2224 +16,2011 @@
...
-            FieldEditLine edit_line_view = editLines.get(i);
-            mCustomViewIds.add(edit_line_view.getId());
-            FieldEditText newEditText = edit_line_view.getEditText();
-            newEditText.setImagePasteListener(this::onImagePaste);
-
+        var customTypeface: Typeface? = null
+        val preferences = AnkiDroidApp.getSharedPrefs(baseContext)
+        val customFont = preferences.getString("browserEditorFont", "")
+        if ("" != customFont) {
+            customTypeface = getTypeface(this, customFont!!)
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:1263]: Keep `AnkiFont.`

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 8]
```diff
@@ -16,2224 +16,2011 @@
...
+                showTagsDialog()
+            }
+            KeyEvent.KEYCODE_C -> {
+                if (event.isCtrlPressed && event.isShiftPressed) {
+                    insertCloze(if (event.isAltPressed) AddClozeType.SAME_NUMBER else AddClozeType.INCREMENT_NUMBER)
                     // Anki Desktop warns, but still inserts the cloze
-                    if (!isClozeType()) {
-                        UIUtils.showSimpleSnackbar(this, R.string.note_editor_insert_cloze_no_cloze_note_type, false);
+                    if (!isClozeType) {
+                        showSimpleSnackbar(
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:577]: Keep `UIUtils.`

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 9]
```diff
@@ -16,2224 +16,2011 @@
...
+                        if (entry_lines.size > 1) {
+                            mSourceText!![0] = entry_lines[1]
+                            mSourceText!![1] = entry_lines[0]
+                            mAedictIntent = true
+                            return false
                         }
                     }
-                    UIUtils.showThemedToast(NoteEditor.this, getResources().getString(R.string.intent_aedict_empty),false);
-                    return true;
+                    showThemedToast(
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:680]: Keep `UIUtils.`

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 10]
```diff
@@ -16,2224 +16,2011 @@
...
-    private String[] mSourceText;
-
-    private FieldState mFieldState = FieldState.fromEditor(this);
-
-    private Toolbar mToolbar;
+    private var mCaller = 0
+    private var mEditFields: LinkedList<FieldEditText?>? = null
+    private var mProgressDialog: MaterialDialog? = null
+    private var mSourceText: Array<String?>? = null
+    private val mFieldState = fromEditor(this)
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:161]: Keep `FieldState.`

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 11]
```diff
@@ -16,2224 +16,2011 @@
...
-    @Nullable
-    private Locale getHintLocaleForField(String name) {
-        JSONObject field = getFieldByName(name);
-        if (field == null) {
-            return null;
-        }
-        return LanguageHintService.getLanguageHintForField(field);
+    private fun getHintLocaleForField(name: String?): Locale? {
+        val field = getFieldByName(name) ?: return null
+        return getLanguageHintForField(field)
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:1539]: Keep `LanguageHintService.`

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 12]
```diff
@@ -16,2224 +16,2011 @@
...
-            default:
-                break;
+            else -> {}
         }
 
         // Note type Selector
-        mNoteTypeSpinner = findViewById(R.id.note_type_spinner);
-        mAllModelIds = NoteTypeSpinnerUtils.setupNoteTypeSpinner(this, mNoteTypeSpinner, col);
+        mNoteTypeSpinner = findViewById(R.id.note_type_spinner)
+        mAllModelIds = setupNoteTypeSpinner(this, mNoteTypeSpinner!!, col)
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:442]: Keep `NoteTypeSpinnerUtils.`

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 13]
```diff
@@ -16,2224 +16,2011 @@
...
             }
-            previous = edit_line_view;
-
-            edit_line_view.setEnableAnimation(animationEnabled());
+            previous = edit_line_view
+            edit_line_view.setEnableAnimation(animationEnabled())
 
             // TODO: Remove the >= M check - one callback works on API 11.
-            if (CompatHelper.getSdkVersion() >= Build.VERSION_CODES.M) {
+            if (sdkVersion >= Build.VERSION_CODES.M) {
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:1291]: Keep `CompatHelper.`

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 14]
```diff
@@ -16,2224 +16,2011 @@
...
                 }
-            });
+            }
         }
 
         // Sets the background color of disabled EditText.
         if (!enabled) {
-            editText.setBackgroundColor(Themes.getColorFromAttr(NoteEditor.this, R.attr.editTextBackgroundColor));
+            editText.setBackgroundColor(
+                getColorFromAttr(
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:1528]: Keep `Themes.`

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 15]
```diff
@@ -16,2224 +16,2011 @@
...
+            }
         }
 
         // Let the user add more buttons (always at the end).
         // Sets the add custom tag icon color.
-        final Drawable drawable = ResourcesCompat.getDrawable(getResources(), R.drawable.ic_add_toolbar_icon, null);
-        drawable.setTint(Themes.getColorFromAttr(NoteEditor.this, R.attr.toolbarIconColor));
-        mToolbar.insertItem(0, drawable, this::displayAddToolbarDialog);
+        val drawable = ResourcesCompat.getDrawable(resources, R.drawable.ic_add_toolbar_icon, null)
+        drawable!!.setTint(getColorFromAttr(this@NoteEditor, R.attr.toolbarIconColor))
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:1724]: Keep `Themes.`

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 16]
```diff
@@ -16,2224 +16,2011 @@
...
-        switch (mCaller) {
-            case CALLER_NO_CALLER:
-                Timber.e("no caller could be identified, closing");
-                finishWithoutAnimation();
-                return;
-
-            case CALLER_REVIEWER:
-                mCurrentEditedCard = AbstractFlashcardViewer.getEditorCard();
+        mToolbar!!.setBackgroundColor(
+            getColorFromAttr(
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:380]: Keep `Themes.`

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 17]
```diff
@@ -16,2224 +16,2011 @@
...
+                this@NoteEditor,
+                R.attr.toolbarBackgroundColor
+            )
+        )
+        mToolbar!!.setIconColor(getColorFromAttr(this@NoteEditor, R.attr.toolbarIconColor))
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:385]: Keep `Themes.`

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 18]
```diff
@@ -16,2224 +16,2011 @@
...
-            newEditText.setCapitalize(prefs.getBoolean("note_editor_capitalize", true));
-
-            ImageButton mediaButton = edit_line_view.getMediaButton();
-            ImageButton toggleStickyButton = edit_line_view.getToggleSticky();
+            newEditText.setCapitalize(prefs.getBoolean("note_editor_capitalize", true))
+            val mediaButton = edit_line_view.mediaButton
+            val toggleStickyButton = edit_line_view.toggleSticky
             // Load icons from attributes
-            int[] icons = Themes.getResFromAttr(this, new int[] { R.attr.attachFileImage, R.attr.upDownImage, R.attr.toggleStickyImage});
+            val icons = getResFromAttr(
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:1308]: Keep `Themes.`

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 19]
```diff
@@ -16,2224 +16,2011 @@
...
+                    Timber.w("error in obtaining templates from model %s", mAllModelIds!![pos])
+                    return
+                }
                 // Initialize mapping between fields of old model -> new model
-                int itemsLength = mEditorNote.items().length;
-                mModelChangeFieldMap = HashUtil.HashMapInit(itemsLength);
-                for (int i=0; i < itemsLength; i++) {
-                    mModelChangeFieldMap.put(i, i);
+                val itemsLength = mEditorNote!!.items().size
+                mModelChangeFieldMap = HashMapInit(itemsLength)
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:1990]: Keep `HashUtil.`

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 20]
```diff
@@ -16,2224 +16,2011 @@
...
-            default:
-                break;
+            else -> {}
         }
 
         // 7573: Ctrl+Shift+[Num] to select a field
-        if (event.isCtrlPressed() && event.isShiftPressed() && KeyUtils.isDigit(event)) {
-            int digit = KeyUtils.getDigit(event);
+        if (event.isCtrlPressed && event.isShiftPressed && isDigit(event)) {
+            val digit = getDigit(event)
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:596]: Keep `KeyUtils`

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 21]
```diff
@@ -16,2224 +16,2011 @@
...
+            popup.setOnMenuItemClickListener { item: MenuItem ->
                 // Get menu item id
-                int idx = item.getItemId();
-                Timber.i("NoteEditor:: User chose to remap to old field %d", idx);
+                val idx = item.itemId
+                Timber.i("NoteEditor:: User chose to remap to old field %d", idx)
                 // Retrieve any existing mappings between newFieldIndex and idx
-                Integer previousMapping = MapUtil.getKeyByValue(mModelChangeFieldMap, newFieldIndex);
-                Integer mappingConflict = mModelChangeFieldMap.get(idx);
+                val previousMapping = getKeyByValue(mModelChangeFieldMap!!, newFieldIndex)
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:1463]: Keep `MapUtil.`

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt [hunk 22]
```diff
@@ -16,2224 +16,2011 @@
...
 
+    // Note: We're not being accurate here - the initial value isn't actually what's supplied in the layout.xml
+    // So a value of 18sp in the XML won't be 18sp on the TextView, but it's close enough.
+    // Values are setFontSize are whole when returned.
+    private val editTextFontSize: String
+        get() {
+            // Note: We're not being accurate here - the initial value isn't actually what's supplied in the layout.xml
+            // So a value of 18sp in the XML won't be 18sp on the TextView, but it's close enough.
+            // Values are setFontSize are whole when returned.
+            val sp = getTextSizeSp(mEditFields!!.first!!)
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.kt:975]: Keep `getTextSizeSp`

**Review by BrayanDSO — COMMENTED** [review]
Some import nits. 
Kotlin conversion of static imports is odd to me. Commonly it removes the java static imports and add some random new ones.

When I do a Kotlin Migration, I normally delete all the static methods on Kotlin then I copy the previous ones from Java. This makes it easier for myself to keep the imports correct. I hope it helps you.

**Comment by david-allison** [comment]
@BrayanDSO Thank you so much for the thorough review

**Review by mikehardy — APPROVED** [review]
Holy heck that thing is big :exploding_head:  Let's go!

**Pull request merged**